### PR TITLE
use cuhpx for reordering on gpu

### PR DESCRIPTION
**Pull request by nbren12**

### CODE REVIEW

File: earth2grid/healpix.py [hunk 1]
```diff
@@ -91,10 +96,57 @@ def pad(x: torch.Tensor, padding: int) -> torch.Tensor:
...
+    def to_nest_cuda(self, x: torch.Tensor):
+        if self == PixelOrder.RING:
+            return _apply_cuhpx_remap(cuhpx.ring2nest, x)
+        elif self == PixelOrder.NEST:
+            return x
+
+    def to_xy_cuda(self, x: torch.Tensor, dest: "XY"):
+        if self == PixelOrder.RING:
+            return _apply_cuhpx_remap(cuhpx.ring2flat, x, clockwise=dest.clockwise, origin=dest.origin.name)
+        elif self == PixelOrder.NEST:
```
Comment (akshaysubr) [inline earth2grid/healpix.py:147]: Why no check for `XY`? You could use `xy2xy` which would be nice for arbitrary `XY to `HEALPIX_PAD_XY` and back if needed right?

File: earth2grid/healpix.py [hunk 2]
```diff
@@ -91,10 +96,57 @@ def pad(x: torch.Tensor, padding: int) -> torch.Tensor:
...
+
+def nside2level(nside: int):
+    return int(math.log2(nside))
+
+
 class PixelOrder(Enum):
     RING = 0
     NEST = 1
 
+    def reorder_from_cuda(self, x, src: "PixelOrderT"):
```
Comment (akshaysubr) [inline earth2grid/healpix.py:126]: Is it normal for Enum classes to have compute methods like this?

File: earth2grid/healpix.py [hunk 3]
```diff
@@ -262,13 +344,19 @@ def get_bilinear_regridder_to(self, lat: np.ndarray, lon: np.ndarray):
     def approximate_grid_length_meters(self):
         return approx_grid_length_meters(self._nside())
 
-    def reorder(self, order: PixelOrderT, x: torch.Tensor) -> torch.Tensor:
-        """Rorder the pixels of ``x`` to have ``order``"""
+    def _reorder_cpu(self, x: torch.Tensor, order: PixelOrderT):
```
Comment (akshaysubr) [inline earth2grid/healpix.py:347]: Could `reorder_from_cuda` and `_reorder_cpu` be in the same place? Would be nicer for future maintainability.

**Review by akshaysubr — APPROVED** [review]
Looks good to me, just added a few comments on code structure and one comment on additional functionality that might be good to add.

File: earth2grid/healpix.py (hunk 1 above)
Comment (nbren12) [reply earth2grid/healpix.py:147]: `XY` is a separate class, so has its own versions of these. These methods aren't intended to be public. But are merely implementing a double dispatch pattern to simplify the logic in `earth2grid.healpix.reorder` and `earthgrid.healpix.Grid.reorder`. The cpu/gpu logic is also handled in those two functions.

File: earth2grid/healpix.py (hunk 2 above)
Comment (nbren12) [reply earth2grid/healpix.py:126]: it works, and is tested. not public API. ultimately, I will probably refactor PixelOrder to be a class hierarchy.

File: earth2grid/healpix.py (hunk 3 above)
Comment (nbren12) [reply earth2grid/healpix.py:347]: The cpu and gpu reordering share almost no logic, so I split them at a higher level in the code.

**Comment by nbren12** [comment]
Thanks Akshay. please note that all the to_xy_cuda methods are following a double dispatch pattern and not intended to be public.

**Comment by nbren12** [comment]
I chose OO-style double dispatch to avoid writing a long reorder function with many if-statements.

**Comment by akshaysubr** [comment]
Ah, sorry, I missed the `XY.to_xy_cuda` method which already has the functionality that I incorrectly thought was missing. Looks good to merge now!

**Pull request merged**